Add border subregion

mrbuds opened the pull request:
# Description

Add border subregion for: icon, aurabar, texture, progresstexture

TODO:
- ~~migrate aurabar border~~
- add backdrop options ? (currently it's a feature loss for aurabar) => i'd prefer not, and if we really wanted it i think it would be better as a separate subregion
- ~~use it for dynamic group ?~~ no, maybe later
- ~~frameLevel~~ => let user decide with subregion order seems good to me
- ~~Make "Border X" number independent from other subregion type~~

Fixes #1138, #1558
## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

Stanzilla commented:
Do we want people adding more than one border?

InfusOnWoW commented:
I think we probably should limit that to one. As to framelevels, which is on your todo list:
* We either just make use of the automatic frame level determination that is currently implemented, meaning users can choose how to stack their sub elements (which doesn't sound too useful for texts, but might become a bit more interesting with more types)
* We add a flag to the sub region type, that this sub region doesn't take part in the automatic stacking, and set a fixed frame level offset based on region type

mrbuds commented:
In current state, group & dynamicgroup are not using the border subregion. For them, having backdrop can make sense contrary to aurabar/icon/texture/... I like how border settings are displayed for groups & dynamicgroup.

I'm not happy with the add border subregion button, i wish it could be shown like it's done with groups (minus backdrop) and still be a subregion for the code structure improvement it gives.

emptyrivers commented on WeakAuras/SubRegionTypes/SubBorder.lua:
```diff
@@ -0,0 +1,88 @@
+if not WeakAuras.IsCorrectVersion() then return end
```
Oh, no no no no. We aren’t doing this, please. I relented when @InfusOnWoW named the text subregion filed as SubText, but I am *not* pleased with the idea of naming every damn file in this directory SubSomething. That’s really dumb.

It’s a border, so just call it Border.lua

Stanzilla replied:
I agree :D

emptyrivers commented:
I don’t understand why we wouldn’t want multiple borders. In not too long, I anticipate that we’ll have the ability to add multiple textures to the same aura. It is not unreasonable that a user would want a border on each texture.

mrbuds commented:
I renamed the files.

So big question:
- Allow only 1 border per aura, show it as a subsection of "Icon Settings" / "Progress Bars Settings" / etc..
- Allow only 1 border per aura, add it with the "Add Border" button in the "Extra Elements" section (how it's currently done)
- Unlimited number of borders per aura, added with the "Add Border" button in the "Extra Elements" section

Stanzilla commented:
I'm not sure, as a first step I would leave it like is currently done.

mrbuds commented:
After discussion on discord, i have re-added possibility to add infinite number of borders.
And for AuraBar regiontype i have added more anchor option https://i.imgur.com/eoHnbmq.png

InfusOnWoW commented on WeakAuras/WeakAuras.lua:
```diff
@@ -3412,6 +3414,47 @@ function WeakAuras.Modernize(data)
+      data.borderOffset = nil
+      data.borderInset = nil
+      data.borderSize = nil
+      data.backdropColor = nil
```
I wonder whether we should leave the backdrop related data in aurabar. If there are users that make use of backdrops, we can readd backdrops and have the right data

InfusOnWoW commented on WeakAurasOptions/SubRegionOptions/Border.lua:
```diff
@@ -0,0 +1,118 @@
+
+local screenWidth, screenHeight = math.ceil(GetScreenWidth() / 20) * 20, math.ceil(GetScreenHeight() / 20) * 20;
+
+local function DeleteSubRegion(data, index, regionType)
```
These functions shouldn't be duplicated, we need to find a common place for them.

Also I'm adding duplicate in my PR.

InfusOnWoW commented:
Overall looks pretty good . One suggestion, and one thing that should be cleaned up. 

I'll have to consider the frame level thing a bit more and figure out to which levels the aurabar assigns, and what we did before.

mrbuds commented:
I copied what you did in https://github.com/WeakAuras/WeakAuras2/pull/1586/commits/dfc007dfd0dc104f7a375eff0b45ff50cc2afa2e so they will conflict

Stanzilla commented:
Superseded by https://github.com/WeakAuras/WeakAuras2/pull/1588
